File: Day 9/Part 2/solve.py
# Do not use this code, use the Java Code


import sys

def solve(input_str):
    lines = input_str.strip().splitlines()
    points = [tuple(map(int, line.split(','))) for line in lines]
    n = len(points)

    # Close the polygon
    polygon = points + [points[0]]

    # Unique sorted x and y coordinates
    xs = sorted({p[0] for p in points})
    ys = sorted({p[1] for p in points})
    nx = len(xs)
    ny = len(ys)

    # Mappings from coordinate to index
    x_to_idx = {x: i for i, x in enumerate(xs)}
    y_to_idx = {y: i for i, y in enumerate(ys)}

    # Build vertical edges (for ray casting)
    vertical_edges = []
    for i in range(n):
        (x1, y1), (x2, y2) = polygon[i], polygon[i+1]
        if x1 == x2:
            if y1 > y2:
                y1, y2 = y2, y1
            vertical_edges.append((x1, y1, y2))

    # For each interval between consecutive y's, collect x-coordinates of vertical edges that span it
    interval_xs = [[] for _ in range(ny-1)]
    for x, y1, y2 in vertical_edges:
        j1 = y_to_idx[y1]
        j2 = y_to_idx[y2]
        for j in range(j1, j2):   # intervals j with ys[j] < ys[j+1] inside [y1,y2]
            interval_xs[j].append(x)

    for j in range(ny-1):
        interval_xs[j].sort()

    # Determine which grid cells are entirely inside the polygon
    # Cell (i,j) corresponds to x in [xs[i], xs[i+1]] and y in [ys[j], ys[j+1]]
    allowed = [[0]*(ny-1) for _ in range(nx-1)]
    for i in range(nx-1):
        xc = (xs[i] + xs[i+1]) / 2.0
        for j in range(ny-1):
            # Count how many vertical edges to the right of xc in this y-interval
            lst = interval_xs[j]
            # Use binary search to find first index with x > xc
            lo, hi = 0, len(lst)
            while lo < hi:
                mid = (lo + hi) // 2
                if lst[mid] > xc:
                    hi = mid
                else:
                    lo = mid + 1
            cnt = len(lst) - lo
            allowed[i][j] = 1 if cnt % 2 == 1 else 0

    # 2D prefix sums for allowed cells
    prefix = [[0]*(ny) for _ in range(nx)]
    for i in range(1, nx):
        for j in range(1, ny):
            prefix[i][j] = allowed[i-1][j-1] + prefix[i-1][j] + prefix[i][j-1] - prefix[i-1][j-1]

    # Check all pairs of red points
    max_area = 0
    for a in range(n):
        x1, y1 = points[a]
        for b in range(a+1, n):
            x2, y2 = points[b]
            if x1 == x2 or y1 == y2:
                continue   # skip degenerate rectangles (lines)
            xl, xr = min(x1, x2), max(x1, x2)
            yb, yt = min(y1, y2), max(y1, y2)
            i1 = x_to_idx[xl]
            i2 = x_to_idx[xr]
            j1 = y_to_idx[yb]
            j2 = y_to_idx[yt]
            if i2 == i1 or j2 == j1:
                continue   # no cell area
            # Cells from i1 to i2-1, j1 to j2-1
            cell_area = (i2 - i1) * (j2 - j1)
            total_allowed = (prefix[i2][j2] - prefix[i1][j2] -
                             prefix[i2][j1] + prefix[i1][j1])
            if total_allowed == cell_area:
                area = (xr - xl + 1) * (yt - yb + 1)
                if area > max_area:
                    max_area = area

    return max_area

if __name__ == "__main__":
    # Read from file if provided, otherwise from stdin
    if len(sys.argv) > 1:
        with open(sys.argv[1], "r") as f:
            input_str = f.read()
    else:
        input_str = sys.stdin.read()
    print(solve(input_str))

File: Day 9/Part 2/solve.java
import java.util.*;
import java.io.*;

public class solve {
    static class Point {
        long x, y;
        Point(long x, long y) {
            this.x = x;
            this.y = y;
        }
    }
    
    public static void main(String[] args) throws IOException {
        // Read from input.txt file in the same directory
        BufferedReader br = new BufferedReader(new FileReader("input.txt"));
        List<Point> points = new ArrayList<>();
        String line;
        while ((line = br.readLine()) != null && !line.isEmpty()) {
            String[] parts = line.split(",");
            long x = Long.parseLong(parts[0]);
            long y = Long.parseLong(parts[1]);
            points.add(new Point(x, y));
        }
        br.close();
        
        System.out.println(solveProblem(points));
    }
    
    static long solveProblem(List<Point> points) {
        int n = points.size();
        if (n == 0) return 0;
        
        // Close the polygon
        List<Point> polygon = new ArrayList<>(points);
        polygon.add(points.get(0));
        
        // Collect unique x and y coordinates
        TreeSet<Long> xSet = new TreeSet<>();
        TreeSet<Long> ySet = new TreeSet<>();
        for (Point p : points) {
            xSet.add(p.x);
            ySet.add(p.y);
        }
        
        List<Long> xs = new ArrayList<>(xSet);
        List<Long> ys = new ArrayList<>(ySet);
        int nx = xs.size();
        int ny = ys.size();
        
        // Mappings from coordinate to index
        Map<Long, Integer> xToIdx = new HashMap<>();
        Map<Long, Integer> yToIdx = new HashMap<>();
        for (int i = 0; i < nx; i++) xToIdx.put(xs.get(i), i);
        for (int i = 0; i < ny; i++) yToIdx.put(ys.get(i), i);
        
        // Build vertical edges
        List<Edge> verticalEdges = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            Point p1 = polygon.get(i);
            Point p2 = polygon.get(i + 1);
            if (p1.x == p2.x) {
                long y1 = p1.y;
                long y2 = p2.y;
                if (y1 > y2) {
                    long temp = y1;
                    y1 = y2;
                    y2 = temp;
                }
                verticalEdges.add(new Edge(p1.x, y1, y2));
            }
        }
        
        // For each y interval, collect x-coordinates of vertical edges that span it
        List<List<Long>> intervalXs = new ArrayList<>(ny - 1);
        for (int i = 0; i < ny - 1; i++) {
            intervalXs.add(new ArrayList<>());
        }
        
        for (Edge e : verticalEdges) {
            int j1 = yToIdx.get(e.y1);
            int j2 = yToIdx.get(e.y2);
            for (int j = j1; j < j2; j++) {
                intervalXs.get(j).add(e.x);
            }
        }
        
        for (int j = 0; j < ny - 1; j++) {
            Collections.sort(intervalXs.get(j));
        }
        
        // Determine which grid cells are inside the polygon
        int[][] allowed = new int[nx - 1][ny - 1];
        for (int i = 0; i < nx - 1; i++) {
            double xc = (xs.get(i) + xs.get(i + 1)) / 2.0;
            for (int j = 0; j < ny - 1; j++) {
                List<Long> lst = intervalXs.get(j);
                // Binary search for first index with x > xc
                int lo = 0, hi = lst.size();
                while (lo < hi) {
                    int mid = (lo + hi) / 2;
                    if (lst.get(mid) > xc) {
                        hi = mid;
                    } else {
                        lo = mid + 1;
                    }
                }
                int cnt = lst.size() - lo;
                allowed[i][j] = (cnt % 2 == 1) ? 1 : 0;
            }
        }
        
        // 2D prefix sums
        long[][] prefix = new long[nx][ny];
        for (int i = 1; i < nx; i++) {
            for (int j = 1; j < ny; j++) {
                prefix[i][j] = allowed[i-1][j-1] + prefix[i-1][j] + prefix[i][j-1] - prefix[i-1][j-1];
            }
        }
        
        // Check all pairs of red points
        long maxArea = 0;
        for (int a = 0; a < n; a++) {
            Point p1 = points.get(a);
            long x1 = p1.x, y1 = p1.y;
            for (int b = a + 1; b < n; b++) {
                Point p2 = points.get(b);
                long x2 = p2.x, y2 = p2.y;
                if (x1 == x2 || y1 == y2) continue;
                
                long xl = Math.min(x1, x2), xr = Math.max(x1, x2);
                long yb = Math.min(y1, y2), yt = Math.max(y1, y2);
                
                int i1 = xToIdx.get(xl);
                int i2 = xToIdx.get(xr);
                int j1 = yToIdx.get(yb);
                int j2 = yToIdx.get(yt);
                
                if (i2 == i1 || j2 == j1) continue;
                
                long cellArea = (long)(i2 - i1) * (j2 - j1);
                long totalAllowed = prefix[i2][j2] - prefix[i1][j2] - prefix[i2][j1] + prefix[i1][j1];
                
                if (totalAllowed == cellArea) {
                    long area = (xr - xl + 1) * (yt - yb + 1);
                    if (area > maxArea) {
                        maxArea = area;
                    }
                }
            }
        }
        
        return maxArea;
    }
    
    static class Edge {
        long x, y1, y2;
        Edge(long x, long y1, long y2) {
            this.x = x;
            this.y1 = y1;
            this.y2 = y2;
        }
    }
}